config/lobby/db/004_audit_bans

```diff
@@ -0,0 +1,23 @@
+start transaction;
+
+alter table banned_macs
+  add column mod_username varchar(40),
```

Comment:
Should this be normalized? If a moderator changes name, could be an issue.

Reply:
Mods should never change their name.  A mod is a mod.

Reply:
`username` is the primary key in the `ta_users` table, which would correspond to this field.  So, yes, it's normalized.  I suppose I could create a foreign key relationship between these two tables in the event someone renamed a value in the `ta_users.username` column.  However, existing tables like `banned_usernames` and `muted_usernames` do not have such a precedent.

Note also that, depending on how we set up the constraint, it would prevent deletion of a mod account as long as they were responsible for an active ban.

Reply:
To expand on that if they want to play without mod headaches they just create a new account with the present system which does not give them mod status.  The only reason we have some new mod names is due to the new database.  Yes we added a few new mods but (90%) are same as old.

Reply:
Note also that adding a foreign key constraint will make it impossible to use a dummy value for the `mod_username` column, as @RoiEXLab suggested above.

Reply:
@ssoloff 
>  However, existing tables like banned_usernames and muted_usernames do not have such a precedent.

The only reason I didn't add such a constraint back when I modified the database, was because the toolbox didn't actually check if the username a mod entered would actually exist, resulting in an exception whenever a non existent username is typed in.
On the Mod side however we always know such a player exists. And we could create a dummy mod that "created" all the bans, and advise mods to reban some of their favourite players ^^ so that the entry is being overwritten.
However what would we do in case a mod deletes their account? The logical answer would be that the name simply stays the same, but this wouldn't work with such a constraint.
No better idea though.

Reply:
A compromise idea: I believe it's possible to have the edit username code left join those columns and edit them along with the user

Reply:
"should" mod change their names? probably not. Should we get data integrity problems in the database the fine day when it does happen? Do we actually disallow name change for mods? AFAIK and recall, you can change username, I don't think we restrict that based on mod status (maybe we do, it's a large code base with lots that I don't remember or know). 
-------------------


re: existing precedent and table structure. The existing precedent is not the best. Using a string column for PK does not seem to have been a good move originally, the tables are pretty clearly mostly copy/pasted : |

Not being able to delete a mod account is not necessarily a bad thing. Might make sense for us to mark accounts as 'not active' rather than as deleted. For audit tables - it is can often be that they are denormalized to allow the referenced entities to be deleted but still retain audit data. To another extent a mod is not special here, a mod could lose the mod flag, but we'd still have the FK relationship. For the audit table I think I would agree and avoid the FK relationships potentially. At the same time I'll note that I don't think there is a way to delete users. We will probably never have a real need to delete moderators (will not get anywhere near enough where it makes sense to clean up from a data point of view).

-----------------------

Really answer I think to FK problem is that mod name column would be the name at the time of the ban. A second (missing) table would audit any name changes, so we could trace through that to see what the current name of the moderator is. That name change table would then have a FK to a user. That seems like a bit overkill for now, so a non FK column in audit table for now seems reasonable.